File: MCTabBarController/MCTabBarController/MCTabBar.m
//
//  MCTabBar.m
//  MCTabBarDemo
//
//  Created by chh on 2017/12/18.
//  Copyright © 2017年 Mr.C. All rights reserved.
//  github:https://github.com/Ccalary/MCTabBarController

#import "MCTabBar.h"

#define MCTabBarItemHeight    49.0f

@interface MCTabBar()
@end
@implementation MCTabBar
- (instancetype)init{
    if (self = [super init]){
        [self initView];
    }
    return self;
}

- (void)initView{
    _centerBtn = [UIButton buttonWithType:UIButtonTypeCustom];
    //去除选择时高亮
    _centerBtn.adjustsImageWhenHighlighted = NO;
    [self addSubview:_centerBtn];
}

// 设置layout
- (void)layoutSubviews {
    [super layoutSubviews];
    switch (self.position) {
        case MCTabBarCenterButtonPositionCenter:
            _centerBtn.frame = CGRectMake(([UIScreen mainScreen].bounds.size.width - _centerWidth)/2.0, (MCTabBarItemHeight - _centerHeight)/2.0 + self.centerOffsetY, _centerWidth, _centerHeight);
            break;
        case MCTabBarCenterButtonPositionBulge:
             _centerBtn.frame = CGRectMake(([UIScreen mainScreen].bounds.size.width - _centerWidth)/2.0, -_centerHeight/2.0 + self.centerOffsetY, _centerWidth, _centerHeight);
            break;
        default:
            break;
    }
}

- (void)setCenterImage:(UIImage *)centerImage {
    _centerImage = centerImage;
    // 如果设置了宽高则使用设置的大小
    if (self.centerWidth <= 0 && self.centerHeight <= 0){
        //根据图片调整button的位置(默认居中，如果图片中心在tabbar的中间最上部，这个时候由于按钮是有一部分超出tabbar的，所以点击无效，要进行处理)
        _centerWidth = centerImage.size.width;
        _centerHeight = centerImage.size.height;
    }
    [_centerBtn setImage:centerImage forState:UIControlStateNormal];
}

- (void)setCenterSelectedImage:(UIImage *)centerSelectedImage {
    _centerSelectedImage = centerSelectedImage;
    [_centerBtn setImage:centerSelectedImage forState:UIControlStateSelected];
}

//处理超出区域点击无效的问题
- (UIView *)hitTest:(CGPoint)point withEvent:(UIEvent *)event{
    if (self.hidden){
        return [super hitTest:point withEvent:event];
    }else {
        //转换坐标
        CGPoint tempPoint = [self.centerBtn convertPoint:point fromView:self];
        //判断点击的点是否在按钮区域内
        if (CGRectContainsPoint(self.centerBtn.bounds, tempPoint)){
            //返回按钮
            return _centerBtn;
        }else {
            return [super hitTest:point withEvent:event];
        }
    }
}

@end

File: MCTabBarControllerSwift/MCTabBarSwift/MCTabBar.swift
//
//  MCTabBar.swift
//  MCTabBarControllerSwift
//
//  Created by caohouhong on 2018/12/11.
//  Copyright © 2018年 chh. All rights reserved.
//  https://github.com/Ccalary/MCTabBarController

import UIKit

public enum MCTabBarCenterButtonPosition : Int {
    case center // 居中
    case bulge  // 凸出一半
}

class MCTabBar: UITabBar {
    // tabbar高度49固定
    let MCTABBARHEIGHT: CGFloat = 49.0
    
    /**
     中间按钮
     */
    public var centerBtn: UIButton = UIButton(type: .custom)
    
    /**
      中间按钮图片
     */
    var centerImage: UIImage? {
        didSet {
            if (centerWidth <= 0 && centerHeight <= 0){
                centerWidth = centerImage?.size.width ?? 0
                centerHeight = centerImage?.size.height ?? 0
            }
            centerBtn.setImage(centerImage, for: .normal)
        }
    }
    /**
     中间选中图片
     */
    var centerSelectedImage: UIImage? {
        didSet {
            centerBtn.setImage(centerSelectedImage, for: .selected)
        }
    }
    
    /**
       中间按钮偏移量,两种可选，也可以使用centerOffsetY 自定义
     */
    var positon: MCTabBarCenterButtonPosition = .center
    
    /**
     中间按钮的宽和高，默认使用图片宽高
     */
    var centerWidth:CGFloat = 0
    var centerHeight:CGFloat = 0
    /**
     中间按钮偏移量，默认是居中
     */
    var centerOffsetY:CGFloat = 0
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        initView()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        
        switch positon {
        case .center:
            centerBtn.frame = CGRect(x: (UIScreen.main.bounds.size.width - centerWidth)/2.0, y: (MCTABBARHEIGHT - centerHeight)/2.0, width: centerWidth, height: centerHeight)
        case .bulge:
            centerBtn.frame = CGRect(x: (UIScreen.main.bounds.size.width - centerWidth)/2.0, y: -centerHeight/2.0 + centerOffsetY, width: centerWidth, height: centerHeight)
        }
    }
    
    func initView() {
        //去除选择时高亮
        centerBtn.adjustsImageWhenHighlighted = false
        self.addSubview(centerBtn)
    }
    
    //处理超出区域点击无效的问题
    override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
        if !self.isHidden{
            //转换坐标
            let tempPoint = centerBtn.convert(point, from: self)
            //判断点击的点是否在按钮区域内
            if centerBtn.bounds.contains(tempPoint) {
                // 返回按钮
                return centerBtn
            }
        }
        return super.hitTest(point, with: event)
    }
}
